space_invaders.py

```diff
@@ -1 +1,295 @@
+            numbers are converted to ints.
+
+        """
+        pygame.sprite.Sprite.__init__(self)
+        self.image = alien_image
+        self.rect = self.image.get_rect()
+        self.rect = self.rect.move(x-self.rect.left, y-self.rect.top) # move
+                                                    # to (x, y)
+        self.velocity = alien_velocity
+
+    def update(self):
+        """
+        Updates the alien's coordinates.  Reverses the alien's velocity at
+        random or if it hits the edges of the window.
+
+        """
+        self.rect = self.rect.move(self.velocity)
+        if self.rect.left < 0 or self.rect.right > WIDTH: # Alien is outside of
+                                                    # window
+            self.velocity[0] = -self.velocity[0]
+            self.rect = self.rect.move(self.velocity)
+            self.rect = self.rect.move(self.velocity)
+            self.velocity[1] = ALIEN_SPEED # comment out this and following
+                            # line to not have aliens get closer
+            self.rect = self.rect.move(DOWN)
+        if self.image == WINGS_UP: # alternate alien images
+            self.image = WINGS_DOWN
+        else:
+            self.image = WINGS_UP
+
+class Player(pygame.sprite.Sprite):
+
+    def __init__(self):
+        """
+        Initializes player at bottom of window.
+
+        """
+        pygame.sprite.Sprite.__init__(self)
+        self.image = pygame.image.load(PLAYER_IMAGE_FILE).convert()
+        self.rect = self.image.get_rect()
+        self.rect = self.rect.move(0, HEIGHT-self.rect.bottom)
+
+    def move(self, direction):
+        """
+        Moves player sprite in specified direction. Won't let player sprite
+        move outside the window's bounds.
+
+        Parameters:
+        -----------
+        direction: 2-tuple of ints
+            (pixels to the right, pixels to the left)
+        """
+        self.rect = self.rect.move(direction)
+        if self.rect.left < 0: # move player back onto screen
+            self.rect = self.rect.move(-self.rect.left, 0)
+        elif self.rect.right > WIDTH:
+            self.rect = self.rect.move(WIDTH-self.rect.right, 0)
+
+class Laser(pygame.sprite.Sprite):
+
+    def __init__(self, player_rect):
+        """
+        Initializes laser just above player's position.
+
+        Parameters:
+        -----------
+        player_rect: pygame.rect
+            A rect that contains the bounds of the players rectangular image
+
+        """
+        pygame.sprite.Sprite.__init__(self)
+        self.image = pygame.image.load(LASER_IMAGE_FILE).convert()
+        self.rect = self.image.get_rect()
+        x = ((player_rect.left + player_rect.right) / 2) - \
+            ((self.rect.right - self.rect.left) / 2)  # line up bullet with
+                                    # center of player sprite
+        y = player_rect.top + (self.rect.bottom - self.rect.top) # align
```

Comment:
this x and y computation should probably go in a helper function that's documented to explain whats going on